asn-compiler/src/parser/asn/types/int.rs

```diff
@@ -266,6 +273,76 @@ fn parse_actual_params(tokens: &[Token]) -> Result<(Vec<ActualParam>, usize), Er
     Ok((params, consumed))
 }
 
+fn maybe_parse_tag(tokens: &[Token]) -> Result<(Option<Asn1Tag>, usize), Error> {
+    match expect_token(tokens, Token::is_square_begin) {
+        Ok(success) => {
+            if success {
+                match parse_tag(tokens) {
+                    Ok((oid, consumed)) => Ok((Some(oid), consumed)),
+                    Err(e) => Err(e),
+                }
+            } else {
+                Ok((None, 0))
+            }
+        }
+        Err(_) => Ok((None, 0)),
+    }
+}
+
+fn parse_tag(tokens: &[Token]) -> Result<(Asn1Tag, usize), Error> {
+    let mut consumed = 0;
+    if !expect_token(&tokens[consumed..], Token::is_square_begin)? {
+        return Err(unexpected_token!("[", tokens[consumed]));
+    }
+    consumed += 1;
+
+    let class = if expect_one_of_keywords(&tokens[consumed..],
+        &["UNIVERSAL", "APPLICATION", "PRIVATE"])? {
+
+        let token = &tokens[consumed];
+        consumed += 1;
+        let typestr = token.text.as_str();
+        match typestr {
+            "UNIVERSAL" => Asn1TagClass::Universal,
+            "APPLICATION" => Asn1TagClass::Application,
+            "PRIVATE" => Asn1TagClass::Private,
+            _ => return Err(unexpected_token!("", token)),
+        }
+    } else {
+        Asn1TagClass::ContextSpecific
+    };
+
+    if !expect_token(&tokens[consumed..], Token::is_numeric)? {
+        return Err(unexpected_token!("<integer>", tokens[consumed]));
+    }
+    let number_token = &tokens[consumed];
+    let number = number_token
+        .text
+        .parse::<u32>()
+        .map_err(|_| invalid_token!(number_token))?;
+    consumed += 1;
+
+    if !expect_token(&tokens[consumed..], Token::is_square_end)? {
+        return Err(unexpected_token!("]", tokens[consumed]));
+    }
+    consumed += 1;
+
+    let mode = if expect_one_of_keywords(&tokens[consumed..], &["IMPLICIT", "EXPLICIT"])? {
```

Comment:
Similarly this should becom `is_explicit` or not.

Reply:
I think we need three states: EXPLICIT, IMPLICIT or not specified.

If not specified, the module level mode applies and we will resolve that at a later stage.

We can replace "Option<Asn1TagMode>" with "Option<boolean>" but that looks a bit strange.

Reply:
I am sorry! yes you are right - I was thinking with simply a boolean it will be easier, but it will actually make it complicated during resolition, also this is lot more explicit in terms of what is specified in the spec. Guess we can merge this.